string.c

```diff
@@ -6069,6 +6077,259 @@ rb_str_dump(VALUE str)
+	return "\f";
+      case 'v':
+	return "\v";
+      case 'b':
+	return "\b";
+      case 'a':
+	return "\a";
+      case 'e':
+	return "\e";
+      default:
+	UNREACHABLE;
+    }
+}
+
+static int
+undump_after_backslash(VALUE undumped, const char *s, const char *s_end, rb_encoding **penc)
+{
+    unsigned int c, c2;
+    int n, n2, codelen;
+    size_t hexlen;
+    char buf[6];
+    static const rb_encoding *enc_utf8 = NULL;
+
+    c = rb_enc_codepoint_len(s, s_end, &n, *penc);
+    switch (c) {
+      case '\\':
+      case '"':
+	rb_str_cat(undumped, s, n); /* cat itself */
+	n++;
+	break;
+      case 'n':
+      case 'r':
+      case 't':
+      case 'f':
+      case 'v':
+      case 'b':
+      case 'a':
+      case 'e':
+	rb_str_cat(undumped, unescape_ascii(c), n);
+	n++;
+	break;
+      case 'u':
+	if (s+1 >= s_end) {
+	    rb_raise(rb_eArgError, "invalid Unicode escape");
+	}
+	if (enc_utf8 == NULL) enc_utf8 = rb_utf8_encoding();
+	if (*penc != enc_utf8) {
+	    *penc = enc_utf8;
+	    rb_enc_associate(undumped, enc_utf8);
+	    ENC_CODERANGE_CLEAR(undumped);
+	}
+	c2 = rb_enc_codepoint_len(s+1, s_end, NULL, *penc);
+	if (c2 == '{') { /* handle \u{...} form */
+	    const char *hexstr = s + 2;
+	    unsigned int hex;
+	    static const char* const close_brace = "}";
+	    long pos;
+
+	    if (hexstr >= s_end) {
+		rb_raise(rb_eArgError, "unterminated Unicode escape");
+	    }
+	    /* find close brace */
+	    pos = strseq_core(hexstr, s_end, s_end - hexstr, close_brace, 1, 0, *penc);
+	    if (pos < 0) {
+		rb_raise(rb_eArgError, "unterminated Unicode escape");
+	    }
+	    hex = ruby_scan_hex(hexstr, pos, &hexlen);
+	    if (hexlen == 0 || hexlen > 6) {
+		rb_raise(rb_eArgError, "invalid Unicode escape");
+	    }
+	    if (hex > 0x10ffffU) {
+		rb_raise(rb_eArgError, "invalid Unicode codepoint (too large)");
+	    }
+	    if ((hex & 0xfffff800U) == 0xd800U) {
+		rb_raise(rb_eArgError, "invalid Unicode codepoint");
+	    }
+	    codelen = rb_enc_codelen(hex, *penc);
+	    rb_enc_mbcput(hex, buf, *penc);
+	    rb_str_cat(undumped, buf, codelen);
+	    n += rb_strlen_lit("u{}") + hexlen;
+	}
+	else { /* handle \uXXXX form */
+	    unsigned int hex = ruby_scan_hex(s+1, 4, &hexlen);
+	    if (hexlen != 4) {
+		rb_raise(rb_eArgError, "invalid Unicode escape");
+	    }
+	    codelen = rb_enc_codelen(hex, *penc);
+	    rb_enc_mbcput(hex, buf, *penc);
+	    rb_str_cat(undumped, buf, codelen);
+	    n += rb_strlen_lit("uXXXX");
+	}
+	break;
+      case 'x':
+	if (s+1 >= s_end) {
+	    rb_raise(rb_eArgError, "invalid hex escape");
+	}
+	c2 = ruby_scan_hex(s+1, 2, &hexlen);
+	if (hexlen != 2) {
+	    rb_raise(rb_eArgError, "invalid hex escape");
+	}
+	*buf = (char)c2;
+	rb_str_cat(undumped, buf, 1L);
+	n += rb_strlen_lit("xXX");
+	break;
+      case '#':
+	if (s+1 >= s_end) {
+	    rb_str_cat(undumped, s, 1L); /* just '#' */
+	    n++;
+	    break;
+	}
+	n2 = rb_enc_mbclen(s+1, s_end, *penc);
+	if (n2 == 1 && IS_EVSTR(s+1, s_end)) {
```

Comment:
This logic seems not needed.

Reply:
> This logic seems not needed.

Thanks, I may have been a bit confused.  I removed it and added some tests in c6a58b01f1.